Add SIMD instructions to syntax

ngzhian opened the pull request:
Main changes are defining:

- ixx for i8x16, i16x8, i32x4, i64x2
- fxx for f32x4, f64x2
- vxx for ixx and fxx
- the the rest of the SIMD instructions are defined based on these

Unfortunately, the definition of instructions aren't as clean as the existing ones, due to high asymmetry (see https://github.com/WebAssembly/simd/blob/master/proposals/simd/NewOpcodes.md for an overview).

This leads to a lot of specific i8x16, i16x8, i32x4 instructions defined (since i64x2 has to be left out).

I also tweaked the paragraph describing comparisons, to indicate that comparisons can return a boolean, or return a value of the same type, because the SIMD comparisons return S128. I'm not sure if this tweak should stick, we can also classify SIMD comparisons as Binary instructions, but that's confusing too. I guess I don't know if these subcategories are meant to classify instructions by "the kind of things they do" or "the type signature"?

See https://www.ngzhian.com/simd/core/syntax/instructions.html for a preview.

binji commented on document/core/syntax/instructions.rst:
```diff
@@ -71,6 +93,49 @@ These operations closely match respective operations available in hardware.
+     \K{i8x16.}\viunop ~|~
+     \K{i16x8.}\viunop ~|~
+     \K{i32x4.}\viunop \\&&|&
+     \K{i64x2.}\NEG \\&&|&
```
I wonder if it would be clearer to use patterns everywhere and exclude forms that are not allowed; there doesn't seem to be precedent for that, however.

ngzhian replied:
That will be kinda neat, but the asymmetry is pretty bad, such an exclusion list might end up being unwieldy too. Probably a short form for i8x16, i16x8, i32x4 will get us part way there (what would a name for that be small_ixx?)

binji replied:
Yeah, not sure the best name to use. Thoughts on this, @rossberg?

rossberg replied:
In the context of memory access I've called the smaller int types "packed", but does that really fit here? Also, it looks like the irregularity is high enough that we need at least two subsets. And then multiple sets of viunop/vibinop. It may not be worth it.

ngzhian replied:
Yea, packed doesn't fit as well here. We kinda use "pack" to mean packing integers/floats side by side. I agree that it might not be worth it. Let's keep this for now, as I work on this, I'll stare at this and see if I can come up with something better. (And I think future work on the validation/execution portion will help us see if there is a better way to organize this.)

binji commented on document/core/syntax/instructions.rst:
```diff
@@ -105,8 +170,20 @@ These operations closely match respective operations available in hardware.
+     \K{andnot} ~|~
+     \K{or} ~|~
+     \K{xor} \\
+   \production{SIMD unary operator} & \vternop &::=&
```
ternary

binji commented on document/core/syntax/instructions.rst:
```diff
@@ -134,15 +237,25 @@ For each type, several subcategories can be distinguished:
   (the source type of the conversion is the one after the ":math:`\K{\_}`").
 
+.. todo::
+  Do these subcategories have to cover every instruction? E.g. simd shifts don't fit anywhere here, since they take 128-bit int and a 32-bit int.
```
I think it might be clearer to group the SIMD operations separately.

binji replied:
Yeah, I think that would be OK. But I'd prefer separating out comparisons too, since those work pretty differently with SIMD.

rossberg replied:
As noted in my comment above, I think we should not lump SIMD under numeric instructions. They are different beasts. And then we can define different subcategories for them.

ngzhian replied:
Moved SIMD instructions into their own section. Called out "shift" instructions in the subcategory.
I did not add a "convention" section to add `vbinop` as a shorthand for `vibinop` and `vfbinop` yet, I think we can add those in as needed when we get to the validation chapter.

binji commented on document/core/syntax/instructions.rst:
```diff
@@ -154,9 +267,22 @@ Occasionally, it is convenient to group operators together according to the foll
-   \production{binary operator} & \binop &::=& \ibinop ~|~ \fbinop \\
-   \production{test operator} & \testop &::=& \itestop \\
+   \production{binary operator} & \binop &::=&
+   \ibinop ~|~
```
nit: indent

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -47,15 +57,27 @@ These operations closely match respective operations available in hardware.
    \begin{array}{llcl}
    \production{width} & \X{nn}, \X{mm} &::=&
      \K{32} ~|~ \K{64} \\
+   \production{simdwidth} & \X{sss} &::=&
```
This should be `nnn`. But is it even worth introducing it as long as we only have one SIMD type?

ngzhian replied:
Changed to nnn. Probably not worth it, inlined it to be v128 in all the necessary instructions.

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -35,6 +35,16 @@ The following sections group instructions into a number of different categories.
+.. _syntax-vibinop:
+.. _syntax-vsatbinop:
+.. _syntax-vfunop:
+.. _syntax-vfbinop:
```
I think it's nicer if we put all the SIMD instructions into their own new section.

ngzhian replied:
Sg, added a new section: SIMD instructions.

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -71,6 +93,49 @@ These operations closely match respective operations available in hardware.
+     \K{i16x8.}\vibinop ~|~
+     \K{i32x4.}\vibinop \\&&|&
+     \K{i64x2.}\ADD ~|~
+     \K{i64x2.}\SUB \\&&|&
```
I'm confused, why are these not binops?

ngzhian replied:
These are binops, but saying `\K{i64x2.}\vibinop` here will not work, since vibinop includes `min_{sx}` and `max_{sx}`, which is not available on i64x2.

Or do you mean, why these 2 instructions are not in the `binops` part of the "Conventions" section below? If this is the case, I probably missed them. I can add them there.

ngzhian replied:
This list contains all the operations, so there will not be more ops (binops) that will need to be enumerated.
Yes I can definitely take min/max out (not sure what I would call them though, let me come up with something and you can comment on my poor names :) )

2 reasons:
- we didn't see any usage of these during the discussion and benchmarking phase when re-considering i64x2 instructions
- difficult to implement these in a performant manner, especially on ia32 and arm v7

ngzhian replied:
Made `viminmaxop` in cbb383ee5e41e58a43c75544c33c86eab514554c

binji commented on document/core/syntax/instructions.rst:
```diff
@@ -235,15 +410,21 @@ Instructions in this group are concerned with linear :ref:`memory <syntax-mem>`.
      \K{i}\X{nn}\K{.}\STORE\K{8}~\memarg ~|~
      \K{i}\X{nn}\K{.}\STORE\K{16}~\memarg ~|~
      \K{i64.}\STORE\K{32}~\memarg \\&&|&
+     \K{i16x8.}\LOAD\K{8x8}\_\sx ~|~
```
I think we'll want to explain the widening and splat load instructions below too.

ngzhian replied:
Added a paragraph on this, ptal at 1a8af6d.

ngzhian commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,181 @@ Occasionally, it is convenient to group operators together according to the foll
+* *Conversions/Splats*: consume a value of the integer or floating-point type and produce a |V128| result of a specified shape.
+
+.. todo::
+   should comparisons be called out in a separate subcategory? they are essentially the same as binary operations
```
Conversion (in the sense of i32x4 <-> f32x4) can probably go into vbinop (same signature).
I meant this bullet point to be about splats, which are conversions in the same sense as the numeric conversions which convert between types.
I think calling it "Conversions/Splats" confuses things, I'll simply call it "Conversions" (and the instructions that make up the conversions are splats).

ngzhian replied:
Fixed in de59417

ngzhian commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,181 @@ Occasionally, it is convenient to group operators together according to the foll
+
+* *Unary Operations*: consume one |V128| operand and produce one |V128| result.
+
+* *Binary Operations*: consume two |V128| operands and produce one |V128| result.
```
Good point, done!

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,176 @@ Occasionally, it is convenient to group operators together according to the foll
+     \K{max} \\
+   \end{array}
+
+SIMD instructions have a naming convention that
+determines how their operands will be interpreted. An instruction beginning with :math:`\K{i32x4}`
+will interpret its operands as four |i32|, packed side-by-side into a |i128|.
+This prefix, :math:`\K{i32x4}`, is known as the *shape* of the type, and is made up of the underlying element type, :math:`\K{i32}`, and the number of elements or *lanes*, :math:`\K{4}`. Operations are performed lane-wise on each element.
```
```suggestion
SIMD instructions have a naming convention involving a prefix that
determines how their operands will be interpreted.
This prefix describes the *shape* of the operand,
written :math:`t\K{x}N`, and consisting of a packed numeric type :math:`t` and the number of *lanes* :math:`N` of that type.
Operations are performed point-wise on the values of each lane.

.. note::
   For example, the shape :math:`\K{i32x4}` interprets the operand
as four |i32| values, packed into an |i128|.
   The bitwidth of the numeric type :math:`t` times :math:`N` always is 128.
```

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,176 @@ Occasionally, it is convenient to group operators together according to the foll
+
+* *Shifts*: consume a |v128| operand and a |i32| operand, producing one |V128| result.
+
+* *Splats*: consume a value of the integer or floating-point type and produce a |V128| result of a specified shape.
```
```suggestion
* *Splats*: consume a value of numeric type and produce a |V128| result of a specified shape.
```

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,176 @@ Occasionally, it is convenient to group operators together according to the foll
+
+* *Splats*: consume a value of the integer or floating-point type and produce a |V128| result of a specified shape.
+
+* *Extract lanes*: consume a |V128| operand and an immediate byte specifying the lane index and produce a result of the element type.
```
```suggestion
* *Extract lanes*: consume a |V128| operand and return the numeric value in a given lane.
```

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,176 @@ Occasionally, it is convenient to group operators together according to the foll
+
+* *Extract lanes*: consume a |V128| operand and an immediate byte specifying the lane index and produce a result of the element type.
+
+* *Replace lanes*: consume a |V128| operand, an immediate byte specifying the lane index, and a value of the element type, and produce a |V128| result.
```
```suggestion
* *Replace lanes*: consume a |V128| operand and a numeric value for a given lane, and produce a |V128| result.
```

ngzhian commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,176 @@ Occasionally, it is convenient to group operators together according to the foll
+     \K{i32x4.}\vtestop \\&&|&
+     \K{i8x16.}\BITMASK ~|~
+     \K{i16x8.}\BITMASK ~|~
+     \K{i32x4.}\BITMASK \\&&|&
```
What do you think about merging BITMASK into vtestop. Their signatures are the same `v128 -> i32`. Though semantically Bitmask is not a "test".

rossberg replied:
Seems like a stretch, I think I'd leave it separate.

ngzhian commented on document/core/syntax/instructions.rst:
```diff
@@ -170,6 +170,182 @@ Occasionally, it is convenient to group operators together according to the foll
+     \K{v128.}\vsunop \\&&|&
+     \K{v128.}\vsbinop \\&&|&
+     \K{v128.}\vsternop \\&&|&
+     \K{v8x16.}\SHUFFLE ~|~ \K{v8x16.}\SWIZZLE \\&&|&
```
Do we need to talk about `v8x16` below? Or perhaps renaming this to `i8x16`.
The reason `v8x16` was chosen is that the operation doesn't care about the type, just that it is an 8-bit value.
But `v8` isn't really defined anywhere (and is not a "packed numeric type" like the description of *shape* below mentions).

ngzhian replied:
No valid anymore, these have been renamed to use the `i8x16` prefix.

ngzhian commented:
I noticed that some instructions that take immediates were not properly specified:
- v8x16.shuffle
- extract_lane
- replace_lane

Introduced a new production `laneidx`, which is `byte`, and added it to the appropriate instructions. see 18049c7.

ngzhian commented on document/core/syntax/instructions.rst:
```diff
@@ -235,15 +413,21 @@ Instructions in this group are concerned with linear :ref:`memory <syntax-mem>`.
+     \K{i16x8.}\LOAD\K{8x8}\_\sx~\memarg ~|~
+     \K{i32x4.}\LOAD\K{16x4}\_\sx~\memarg ~|~
+     \K{i64x2.}\LOAD\K{32x2}\_\sx~\memarg \\&&|&
+     \X{vxx}\K{.}\LOAD\K{\_splat}~\memarg \\&&|&
```
This is incorrect. vxx represents ixx and fxx (i8x16... f64x2). It should be `v8x16`, `v16x8`, `v32x4`, `v64x2`, might need another production for here.

rossberg replied:
Is there a chance we can just rename all the `vNxM` prefixes to `iNxM`? I understand the reason is that it doesn't matter that they are ints, but the same is true for, say, plain `i32.and` or `i32.load`.

ngzhian replied:
The difference is that `i32.load`  returns `i32` and we have `f32.load` if we want a `f32`, whereas these all return a `v128`. Naming it with a `i` prefix suggests that it is implemented using integer loads, where it is not always the case, and also might cause confusion because there is no `f` counterpart for loading floats.

That said, I don't have a strong preference either way. I can file an issue and see if anyone has comments.
@tlively if you have comments on this re: toolchain or initial rationale for naming loads.

tlively replied:
I actually would prefer using `i` prefixes for these as well because the `v` prefixes are used on few enough instructions that they seem surprising and because they do not correspond to anything in MVP. I recall [discussing](https://github.com/WebAssembly/simd/pull/82#issuecomment-519893850) this for specific instructions in the past, but the `v` prefix was [present](https://github.com/WebAssembly/simd/pull/1/files#diff-28263aca205221ef5f3c6b4a232654c3R281-R285) in the original proposal writeup, so I don't think we've discussed making this change for all instructions before.

There's no toolchain-specific reason to make this change, though.

binji replied:
How much simpler is the definition here if we make the change? I have to admit I'm a bit hesitant to suggest changing the instruction names for the simplicity of the spec document.

ngzhian replied:
> How much simpler is the definition here if we make the change?

This definition of load_splats will become `ixx.load_splat`.

If we stick to `v8x16` et al, we will need to:

1. Define `vX`. Right now we rely on the definition of `iN` to describe the *shape* of SIMD, using `vX` means we need to somehow describe `vX`, but in a way that makes it not part of the syntax
2. After `vX` we can talk about `vXxN`

It won't be a lot more text, but it will be very specialized to these 6 instructions, which (like Thomas says) does not have any analogues in MVP.

We *could* define `f32x4.load_splat` and `f64x2.load_splat`, but have them be "aliases", so share the same opcode and semantics as the integer counterparts...

tlively replied:
I would also be fine just having `i32x4.load_splat` and friends and not having `f32x4.load_splat`. We could also avoid the problem (for these instructions) by renaming them to `v128.load<width>_splat` (and following suit for the load-and-zero instructions).

ngzhian replied:
Yea I quite like the `v128.load<width>_splat` notation we were discussing in #237 , then we can side-step this issue for load_splat, and rename swizzle and shuffle to `i8x16`.

ngzhian replied:
Filed #297 to discuss this more, and see if anyone supports/objects.

ngzhian replied:
#297 hasn't had too many comments outside of the people already involved here, but I think it will be a fine change. The only other change is `v8x16.shuffle` and `v8x16.swizzle`, which I have filed #316 for.

ngzhian replied:
I have optimistically made the naming changes (of load splats, load extends, shuffle, swizze) in the latest commit. With that, we no longer have special `v...` prefix, so no need to define them, and loads all start with `v128`.

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -253,6 +437,8 @@ They all take a *memory immediate* |memarg| that contains an address *offset* an
 Integer loads and stores can optionally specify a *storage size* that is smaller than the :ref:`bit width <syntax-valtype>` of the respective value type.
 In the case of loads, a sign extension mode |sx| is then required to select appropriate behavior.
 
+SIMD loads can specify a shape that is half the :ref:`bit width <syntax-valtype>` of |V128|. Each lane is half its usual size, and the sign extension mode |sx| then specifies how the smaller lane is extended to the larger lane. SIMD loads can be annotated with *splat*, to indicate that only a single lane of the respective shape is loaded, and the result is duplicated to all other lanes.
```
```suggestion
SIMD loads can specify a shape that is half the :ref:`bit width <syntax-valtype>` of |V128|. Each lane is half its usual size, and the sign extension mode |sx| then specifies how the smaller lane is extended to the larger lane. Alternatively, SIMD loads can perform a *splat*, such that only a single lane of the respective shape is loaded, and the result is duplicated to all other lanes.
```

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -210,7 +208,7 @@ SIMD instructions provide basic operations over :ref:`values <syntax-value>` of
      \K{v128.}\vsbinop \\&&|&
      \K{v128.}\vsternop \\&&|&
-     \K{v8x16.}\SHUFFLE~\laneidx^{16} ~|~ \K{v8x16.}\SWIZZLE \\&&|&
+     \K{i8x16.}\SHUFFLE~\laneidx^{16} ~|~ \K{i8x16.}\SWIZZLE \\&&|&
```
Nit: two lines, since they are different ops
